apk: does not conflict when installing packages which touch the same files

### Describe the bug

I have seen things like dnsmasq and dnsmasq-full not conflicting when installed with APK and more recently while working on fixing some packages dependencies with avahi package that APK does not conflicts when installing packages that installs the same files.

```
( 13/165) Installing libavahi-dbus-support (0.8-r10)
( 14/165) Installing libdaemon (0.14-r5)
( 15/165) Installing librt (1.2.5-r5)
( 16/165) Installing avahi-dbus-daemon (0.8-r10)
( 17/165) Installing libavahi-nodbus-support (0.8-r10)
( 18/165) Installing avahi-nodbus-daemon (0.8-r10)
```

OPKG would say:

```
Collected errors:
 * check_data_file_clashes: Package libavahi-dbus-support wants to install file /builder/build_dir/target-mipsel_24kc_musl/root-ramips/usr/lib/libavahi-common.so.3
	But that file is already provided by package  * libavahi-nodbus-support
 * check_data_file_clashes: Package libavahi-dbus-support wants to install file /builder/build_dir/target-mipsel_24kc_musl/root-ramips/usr/lib/libavahi-common.so.3.5.4
	But that file is already provided by package  * libavahi-nodbus-support
 * check_data_file_clashes: Package libavahi-dbus-support wants to install file /builder/build_dir/target-mipsel_24kc_musl/root-ramips/usr/lib/libavahi-core.so.7
	But that file is already provided by package  * libavahi-nodbus-support
 * check_data_file_clashes: Package libavahi-dbus-support wants to install file /builder/build_dir/target-mipsel_24kc_musl/root-ramips/usr/lib/libavahi-core.so.7.1.0
	But that file is already provided by package  * libavahi-nodbus-support
 * opkg_install_cmd: Cannot install package avahi-dbus-daemon.
make[2]: *** [Makefile:234: package_install] Error 255
```

I checked alpine documentation and there should be conflicts. https://wiki.alpinelinux.org/wiki/APKBUILD_Reference#replaces

### OpenWrt version

r31813-971c058cf5

### OpenWrt release

SNAPSHOT

### OpenWrt target/subtarget

ramips/mt7621

### Device

Mi Router 3G

### Image kind

Official downloaded image

### Steps to reproduce

apk add avahi-dbus-daemon avahi-nodbus-daemon

### Actual behaviour

Installs two packages without errors.

### Expected behaviour

Stop installation telling that there are conflicts installing the same files.

### Additional info

A quick test on an out-dated installation: (I also tested on openwrt-firmware-selector for recent SNAPSHOT) 

More weirdness (and possible cause) only one package of the two variants seems to contains the binaries. 
```
root@OpenWrt:~# apk add avahi-dbus-daemon avahi-nodbus-daemon
ERROR: wget: exited with error 8
WARNING: updating and opening https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/kmods/6.6.93-1-3d031212610de6188f9bb0af77a78b73/packages.adb: unexpected end of file
(1/8) Installing libdbus (1.16.2-r3)
libdbus-1.16.2-r3.post-install: Executing script...
(2/8) Installing libexpat (2.7.3-r1)
libexpat-2.7.3-r1.post-install: Executing script...
(3/8) Installing dbus (1.16.2-r3)
dbus-1.16.2-r3.post-install: Executing script...
dbus-1.16.2-r3.post-install: chown: /etc/dbus-1: No such file or directory
(4/8) Installing libavahi-dbus-support (0.8-r10)
libavahi-dbus-support-0.8-r10.post-install: Executing script...
(5/8) Installing libdaemon (0.14-r5)
libdaemon-0.14-r5.post-install: Executing script...
(6/8) Installing avahi-dbus-daemon (0.8-r10)
avahi-dbus-daemon-0.8-r10.post-install: Executing script...
(7/8) Installing libavahi-nodbus-support (0.8-r10)
libavahi-nodbus-support-0.8-r10.post-install: Executing script...
(8/8) Installing avahi-nodbus-daemon (0.8-r10)
avahi-nodbus-daemon-0.8-r10.post-install: Executing script...
OK: 22 MiB in 142 packages
root@OpenWrt:~# apk info -L avahi-dbus-daemon
WARNING: opening from cache https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/kmods/6.6.93-1-3d031212610de6188f9bb0af77a78b73/packages.adb: No such file or directory
avahi-dbus-daemon-0.8-r10 contains:
lib/apk/packages/avahi-dbus-daemon.conffiles
lib/apk/packages/avahi-dbus-daemon.conffiles_static
lib/apk/packages/avahi-dbus-daemon.list
lib/apk/packages/avahi-dbus-daemon.rusers

root@OpenWrt:~# apk info -L avahi-nodbus-daemon
WARNING: opening from cache https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/kmods/6.6.93-1-3d031212610de6188f9bb0af77a78b73/packages.adb: No such file or directory
avahi-nodbus-daemon-0.8-r10 contains:
etc/avahi/avahi-daemon.conf
etc/init.d/avahi-daemon
lib/apk/packages/avahi-nodbus-daemon.conffiles
lib/apk/packages/avahi-nodbus-daemon.conffiles_static
lib/apk/packages/avahi-nodbus-daemon.list
lib/apk/packages/avahi-nodbus-daemon.rusers
usr/sbin/avahi-daemon

root@OpenWrt:~#
root@OpenWrt:~#
root@OpenWrt:~# apk info -L libavahi-dbus-support
WARNING: opening from cache https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/kmods/6.6.93-1-3d031212610de6188f9bb0af77a78b73/packages.adb: No such file or directory
libavahi-dbus-support-0.8-r10 contains:
etc/dbus-1/system.d/avahi-dbus.conf
lib/apk/packages/libavahi-dbus-support.list

root@OpenWrt:~# apk info -L libavahi-nodbus-support
WARNING: opening from cache https://downloads.openwrt.org/snapshots/targets/ramips/mt7621/kmods/6.6.93-1-3d031212610de6188f9bb0af77a78b73/packages.adb: No such file or directory
libavahi-nodbus-support-0.8-r10 contains:
lib/apk/packages/libavahi-nodbus-support.list
usr/lib/libavahi-common.so.3
usr/lib/libavahi-common.so.3.5.4
usr/lib/libavahi-core.so.7
usr/lib/libavahi-core.so.7.1.0

```

### Diffconfig

```text

```

### Terms

- [x] I am reporting an issue for OpenWrt, not an unsupported fork.

--- a/include/package-pack.mk
+++ b/include/package-pack.mk
@@ -199,10 +199,48 @@ endif
 	$(call locked,$(CP) $(PKG_BUILD_DIR)/.pkgdir/$(1)/. $(STAGING_DIR_ROOT)/,root-copy)
 	touch $$@
 
-    Package/$(1)/DEPENDS := $$(call mergelist,$$(foreach dep,$$(filter-out @%,$$(IDEPEND_$(1))),$$(dep)$$(call GetABISuffix,$$(dep))))
+    Package/$(1)/DEPENDS := $$(foreach dep,$$(filter-out @%,$$(IDEPEND_$(1))),$$(dep)$$(call GetABISuffix,$$(dep)))
     ifneq ($$(EXTRA_DEPENDS),)
-      Package/$(1)/DEPENDS := $$(EXTRA_DEPENDS)$$(if $$(Package/$(1)/DEPENDS),$$(comma) $$(Package/$(1)/DEPENDS))
+      ifeq ($(CONFIG_USE_APK),)
+        Package/$(1)/DEPENDS := $$(call mergelist,$$(Package/$(1)/DEPENDS))
+        Package/$(1)/DEPENDS := $$(EXTRA_DEPENDS)$$(if $$(Package/$(1)/DEPENDS),$$(comma) $$(Package/$(1)/DEPENDS))
+      else
+        _SEP := __COMMA_SEP__
+        _SPACE := __SPACE_SEP__
+        _DEPENDS := $$(Package/$(1)/DEPENDS)
+        _EXTRA_DEPENDS_ABI :=
+        _DEP_ITEMS := $$(subst $$(_SEP),$$(space),$$(subst $$(space),$$(_SPACE),$$(subst $$(comma),$$(_SEP),$$(EXTRA_DEPENDS))))
+
+        $$(foreach dep,$$(_DEP_ITEMS), \
+          $$(eval _CUR_DEP := $$(subst $$(_SPACE),$$(space),$$(strip $$(dep)))) \
+          $$(eval _PKG_NAME := $$(word 1,$$(_CUR_DEP))) \
+          $$(if $$(findstring $$(paren_left), $$(_PKG_NAME)), \
+            $$(error "Unsupported extra dependency format: no space before '(': $$(_CUR_DEP)")) \
+          ) \
+          $$(eval _ABI_SUFFIX := $$(call GetABISuffix,$$(_PKG_NAME))) \
+          $$(eval _PKG_NAME_ABI := $$(_PKG_NAME)$$(_ABI_SUFFIX)) \
+          $$(eval _VERSION_CONSTRAINT := $$(word 2,$$(_CUR_DEP))) \
+          $$(if $$(_VERSION_CONSTRAINT), \
+            $$(eval _EXTRA_DEP := $$(_PKG_NAME_ABI) $$(_VERSION_CONSTRAINT)), \
+            $$(error "Extra dependencies must have version constraints. $$(_PKG_NAME) seems to be unversioned.") \
+          ) \
+          $$(if $$(_EXTRA_DEPENDS_ABI), \
+            $$(eval _EXTRA_DEPENDS_ABI := $$(_EXTRA_DEPENDS_ABI)$$(comma)$$(_EXTRA_DEP)), \
+            $$(eval _EXTRA_DEPENDS_ABI := $$(_EXTRA_DEP)) \
+          ) \
+          $$(if $$(_DEPENDS), \
+            $$(eval _DEPENDS := $$(filter-out $$(_PKG_NAME_ABI),$$(_DEPENDS))) \
+          ) \
+        )
+
+        _DEPENDS := $$(call mergelist,$$(_DEPENDS))
+        Package/$(1)/DEPENDS := $$(_EXTRA_DEPENDS_ABI)$$(if $$(_DEPENDS),$$(comma) $$(_DEPENDS))
+      endif
+    else
+      Package/$(1)/DEPENDS := $$(call mergelist,$$(Package/$(1)/DEPENDS))
     endif
+
+    $$(info $(1) fused dependencies: $$(Package/$(1)/DEPENDS))
 
 $(_define) Package/$(1)/CONTROL
 Package: $(1)$$(ABIV_$(1))
@@ -320,7 +358,7 @@ else
 		echo "[ -s "\$$$${IPKG_INSTROOT}/lib/functions.sh" ] || exit 0"; \
 		echo ". \$$$${IPKG_INSTROOT}/lib/functions.sh"; \
 		echo 'export root="$$$${IPKG_INSTROOT}"'; \
-		echo 'export pkgname="$(1)"'; \
+		echo 'export pkgname="$(1)$$(ABIV_$(1))"'; \
 		echo "add_group_and_user"; \
 		echo "default_postinst"; \
 		[ ! -f $$(ADIR_$(1))/postinst-pkg ] || sed -z 's/^\s*#!/#!/' "$$(ADIR_$(1))/postinst-pkg"; \
@@ -345,34 +383,34 @@ else
 		echo "[ -s "\$$$${IPKG_INSTROOT}/lib/functions.sh" ] || exit 0"; \
 		echo ". \$$$${IPKG_INSTROOT}/lib/functions.sh"; \
 		echo 'export root="$$$${IPKG_INSTROOT}"'; \
-		echo 'export pkgname="$(1)"'; \
+		echo 'export pkgname="$(1)$$(ABIV_$(1))"'; \
 		echo "default_prerm"; \
 		[ ! -f $$(ADIR_$(1))/prerm-pkg ] || sed -z 's/^\s*#!/#!/' "$$(ADIR_$(1))/prerm-pkg"; \
 	) > $$(ADIR_$(1))/pre-deinstall;
 
 	[ ! -f $$(ADIR_$(1))/postrm ] || sed -zi 's/^\s*#!/#!/' "$$(ADIR_$(1))/postrm";
 
-	if [ -n "$(USERID)" ]; then echo $(USERID) > $$(IDIR_$(1))/lib/apk/packages/$(1).rusers; fi;
-	if [ -n "$(ALTERNATIVES)" ]; then echo $(ALTERNATIVES) > $$(IDIR_$(1))/lib/apk/packages/$(1).alternatives; fi;
-	(cd $$(IDIR_$(1)) && find . -type f,l -printf "/%P\n" | sort > $(TMP_DIR)/$(1).list && mv $(TMP_DIR)/$(1).list $$(IDIR_$(1))/lib/apk/packages/$(1).list)
+	if [ -n "$(USERID)" ]; then echo $(USERID) > $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).rusers; fi;
+	if [ -n "$(ALTERNATIVES)" ]; then echo $(ALTERNATIVES) > $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).alternatives; fi;
+	(cd $$(IDIR_$(1)) && find . -type f,l -printf "/%P\n" | sort > $(TMP_DIR)/$(1).list && mv $(TMP_DIR)/$(1).list $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).list)
 	# Move conffiles to IDIR and build conffiles_static with csums
 	if [ -f $$(ADIR_$(1))/conffiles ]; then \
-		mv -f $$(ADIR_$(1))/conffiles $$(IDIR_$(1))/lib/apk/packages/$(1).conffiles; \
-		for file in $$$$(cat $$(IDIR_$(1))/lib/apk/packages/$(1).conffiles); do \
+		mv -f $$(ADIR_$(1))/conffiles $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).conffiles; \
+		for file in $$$$(cat $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).conffiles); do \
 			[ -f $$(IDIR_$(1))/$$$$file ] || continue; \
 			csum=$$$$($(MKHASH) sha256 $$(IDIR_$(1))/$$$$file); \
-			echo $$$$file $$$$csum >> $$(IDIR_$(1))/lib/apk/packages/$(1).conffiles_static; \
+			echo $$$$file $$$$csum >> $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).conffiles_static; \
 		done; \
 	fi
 
 	# Some package (base-files) manually append stuff to conffiles
 	# Append stuff from it and delete the CONTROL directory since everything else should be migrated
 	if [ -f $$(IDIR_$(1))/CONTROL/conffiles ]; then \
-		echo $$$$(IDIR_$(1))/CONTROL/conffiles >> $$(IDIR_$(1))/lib/apk/packages/$(1).conffiles; \
+		echo $$$$(IDIR_$(1))/CONTROL/conffiles >> $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).conffiles; \
 		for file in $$$$(cat $$(IDIR_$(1))/CONTROL/conffiles); do \
 			[ -f $$(IDIR_$(1))/$$$$file ] || continue; \
 			csum=$$$$($(MKHASH) sha256 $$(IDIR_$(1))/$$$$file); \
-			echo $$$$file $$$$csum >> $$(IDIR_$(1))/lib/apk/packages/$(1).conffiles_static; \
+			echo $$$$file $$$$csum >> $$(IDIR_$(1))/lib/apk/packages/$(1)$$(ABIV_$(1)).conffiles_static; \
 		done; \
 		rm -rf $$(IDIR_$(1))/CONTROL/conffiles; \
 	fi
@@ -394,16 +432,13 @@ else
 	  --info "origin:$(SOURCE)" \
 	  --info "url:$(URL)" \
 	  --info "maintainer:$(MAINTAINER)" \
-	  --info "provides:$$(foreach prov,\
-			$$(filter-out $(1)$$(ABIV_$(1)), \
-			$(PROVIDES)$$(if $$(ABIV_$(1)), \
-				$(1)=$(VERSION) $(foreach provide, \
-					$(PROVIDES), \
-					$(provide)$$(ABIV_$(1))=$(VERSION) \
-				) \
-			) \
-		), \
-		$$(prov) )" \
+	  --info "provides:$$(if $$(ABIV_$(1)), \
+	    $(1) $(foreach provide,$(PROVIDES), $(provide)$$(ABIV_$(1))=$(VERSION)), \
+	    $(if $(ALTERNATIVES), \
+	      $(PROVIDES), \
+	      $(foreach provide,$(PROVIDES), $(provide)=$(VERSION)) \
+	    ) \
+	  )" \
 	  $(if $(DEFAULT_VARIANT),--info "provider-priority:100",$(if $(PROVIDES),--info "provider-priority:1")) \
 	  $$(APK_SCRIPTS_$(1)) \
 	  --info "depends:$$(foreach depends,$$(subst $$(comma),$$(space),$$(subst $$(space),,$$(subst $$(paren_right),,$$(subst $$(paren_left),,$$(Package/$(1)/DEPENDS))))),$$(depends))" \

